lib/remote_node.go

```diff
@@ -0,0 +1,468 @@
+	}
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+	vMeta.minTxFeeRateNanosPerKB = verMsg.MinFeeRateNanosPerKB
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	vMeta.timeOffsetSecs = verMsg.TstampSecs - time.Now().Unix()
+	vMeta.StartingBlockHeight = verMsg.StartBlockHeight
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
+
+	if !rn.peer.IsOutbound() {
+		// Respond to the version message if this is an inbound peer.
+		rn.sendVersionMessage(responseNonce)
+	}
+	// After sending and receiving a compatible version, complete the
+	// negotiation by sending and receiving a verack message.
+	verackTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+	vMeta.verackTimeExpected = &verackTimeExpected
+	if err := rn.sendVerack(); err != nil {
+		glog.Errorf("RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
+		rn.Disconnect()
+		return
+	}
+}
+
+func (rn *RemoteNode) sendVerack() error {
+	verackMsg, err := rn.newVerackMessage()
+	if err != nil {
+		return err
+	}
+
+	if err := rn.cmgr.SendMessage(verackMsg, rn.peer.ID); err != nil {
+		glog.Errorf("RemoteNode.SendVerack: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
+		return err
+	}
+	return nil
+}
+
+func (rn *RemoteNode) newVerackMessage() (*MsgDeSoVerack, error) {
+	verack := NewMessage(MsgTypeVerack).(*MsgDeSoVerack)
+	vMeta := rn.GetHandshakeMetadata()
+
+	// Include the nonce we received in the peer's version message
+	verack.NonceReceived = vMeta.versionNoncesReceived
+	if vMeta.negotiatedProtocolVersion == ProtocolVersion2 {
+		var err error
+		verack.Version = VerackVersion1
+		verack.NonceSent = vMeta.versionNoncesSent
+		verack.PublicKey = rn.keystore.GetSigner().GetPublicKey()
+		tstampMicro := uint64(time.Now().UnixMicro())
+		verack.Signature, err = rn.keystore.GetSigner().SignPoSValidatorHandshake(verack.NonceSent, verack.NonceReceived, tstampMicro)
+		if err != nil {
+			return nil, fmt.Errorf("RemoteNode.newVerackMessage: Problem signing verack message: %v", err)
+		}
+	}
+	return verack, nil
+}
+
+func (rn *RemoteNode) HandleVerackMessage(vrkMsg *MsgDeSoVerack) {
+	vMeta := rn.GetHandshakeMetadata()
+	nonceReceived := vMeta.versionNoncesReceived
+	nonceSent := vMeta.versionNoncesSent
+
+	if vMeta.verackTimeExpected != nil && vMeta.verackTimeExpected.Before(time.Now()) {
+		glog.V(1).Infof("RemoteNode.HandleVerackMessage: Requesting PeerDisconnect for id: (%v) "+
+			"verack timeout. Time expected: %v, now: %v", rn.peer.ID, vMeta.verackTimeExpected.UnixMicro(), time.Now().UnixMicro())
+		rn.cmgr.CloseConnection(rn.peer.ID)
+		return
+	}
+	// If the verack message has a nonce unseen for us, then request peer disconnect.
+	if vrkMsg.NonceReceived != nonceSent && vrkMsg.NonceReceived != nonceReceived {
```

Comment:
This condition is confusing to me, but basically the `NonceReceived` must be either the nonceReceived or the nonceSent? or do we need to be checking `vrkMsg.NonceSent != nonceSent` and `vrk.NonceReceived != nonceReceived`?

Reply:
Should be fixed now.